src/v/wasm/transform_module.cc

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "transform_module.h"
+
+#include "bytes/iobuf_parser.h"
+#include "model/compression.h"
+#include "model/record.h"
+#include "model/record_batch_types.h"
+#include "model/timestamp.h"
+#include "utils/named_type.h"
+#include "utils/vint.h"
+#include "vassert.h"
+#include "wasm/ffi.h"
+#include "wasm/logger.h"
+
+#include <algorithm>
+#include <exception>
+#include <optional>
+#include <stdexcept>
+#include <vector>
+namespace wasm {
+
+model::record_batch transform_module::for_each_record(
+  const model::record_batch* input,
+  ss::noncopyable_function<void(wasm_call_params)> func) {
+    vassert(
+      input->header().attrs.compression() == model::compression::none,
+      "wasm transforms expect uncompressed batches");
```

Comment:
i assume we are checking for this condition somewhere else too in a way where we can reject it gracefully?